### ISSUE TITLE
Significant drop in accuracy with converted BERT models

### ISSUE THREAD

**Issue by alexriggio** [OpenNMT/CTranslate2#1286]
Running the BERT [example code](https://opennmt.net/CTranslate2/guides/transformers.html) in the docs I am seeing the accuracy drop from 97.56% to 89.3% when performing inference on 5000 samples from the [yelp_polarity test dataset](https://huggingface.co/datasets/yelp_polarity). INT8 quantization does not further reduce accuracy compared to the already converted model. 

Original model: correct=4878,  acc=97.56,  loss=0.107
Converted model: correct=4465, acc=89.3, loss=0.26
Converted model with INT8 Quantization: correct=4463, acc=89.26, loss=0.259

This was run on an NVIDIA A6000 GPU.

**Comment by guillaumekln** [comment]
Can you share the code you are using the compute and compare the accuracy?

**Comment by alexriggio** [comment]
Switched over to the huggingface trainer/evaluation to reduce the chance of errors in my implementation, but still look to be getting the same results.

```
!ct2-transformers-converter --model textattack/bert-base-uncased-yelp-polarity --output_dir bert-base-uncased-yelp-polarity


import torch
import torch.nn as nn
import numpy as np
from datasets import load_dataset
import evaluate
import transformers
from transformers import TrainingArguments, Trainer, AutoConfig, BertTokenizer
import ctranslate2


class Ct2Bert(torch.nn.Module):
    def __init__(self, encoder, classifier, model_name):
        super().__init__()
        self.config = AutoConfig.from_pretrained(model_name)
        self.encoder = encoder
        self.classifier = classifier

    def forward(self, input_ids, token_type_ids=None, attention_mask=None, labels=None):
        output = self.encoder.forward_batch(input_ids.tolist())
        pooler_output = output.pooler_output

        if device == "cuda":
            pooler_output = torch.as_tensor(pooler_output, device=device)
        else:
            pooler_output = np.array(pooler_output)
            pooler_output = torch.as_tensor(pooler_output)

        logits = self.classifier(pooler_output)
        
        loss = None
        if labels is not None:
            loss_fct = nn.CrossEntropyLoss()
            loss = loss_fct(logits, labels)

        return loss, logits


BATCH_SIZE=1
N_SAMPLES=5000

# data
model_name = "textattack/bert-base-uncased-yelp-polarity"
tokenizer = BertTokenizer.from_pretrained(model_name)

test_dataset = load_dataset("yelp_polarity", split="test")

def tokenize_function(examples):
    return tokenizer(examples["text"], padding="max_length", max_length=512, truncation=True)

test_tokenized = test_dataset.shuffle(seed=42).select(range(N_SAMPLES)).map(tokenize_function, batched=True)


# models
device = "cuda"
encoder = ctranslate2.Encoder("bert-base-uncased-yelp-polarity", device=device)
model_name = "textattack/bert-base-uncased-yelp-polarity"
tokenizer = transformers.AutoTokenizer.from_pretrained(model_name)
original_model = transformers.AutoModelForSequenceClassification.from_pretrained(model_name)
classifier = original_model.classifier
classifier.eval()
classifier.to(device)
ct2_model = Ct2Bert(encoder, classifier, model_name)


# trainers
metric = evaluate.load("accuracy")

def compute_metrics(eval_pred):
    logits, labels = eval_pred
    predictions = np.argmax(logits, axis=-1)
    return metric.compute(predictions=predictions, references=labels)


training_args = TrainingArguments(
    output_dir="test_trainer", 
    do_train = False,
    do_predict = True,
    per_device_eval_batch_size = BATCH_SIZE,   
    dataloader_drop_last = False,    
)


trainer = Trainer(
    model=original_model,
    args=training_args,
    compute_metrics=compute_metrics,
)


ct2_trainer = Trainer(
    model=ct2_model,
    args=training_args,
    compute_metrics=compute_metrics,
)

# predictions
predictions = trainer.predict(test_tokenized)
ct2_predictions = ct2_trainer.predict(test_tokenized)

print(predictions.metrics)
print(ct2_predictions.metrics)
```
```
{'test_loss': 0.1142459362745285, 'test_accuracy': 0.9744, 'test_runtime': 56.1643, 'test_samples_per_second': 89.025, 'test_steps_per_second': 89.025}

{'test_loss': 0.25635623931884766, 'test_accuracy': 0.9, 'test_runtime': 54.2191, 'test_samples_per_second': 92.218, 'test_steps_per_second': 92.218}
```

**Comment by vakkov** [comment]
@guillaumekln Any plans or guidance on how to add a classification head?

**Comment by guillaumekln** [comment]
@vakkov The examples in the [documentation](https://opennmt.net/CTranslate2/guides/transformers.html#bert) and in this issue are already using a classification head. Please create another issue if something is not working or missing.

**Comment by vakkov** [comment]
> @vakkov The examples in the [documentation](https://opennmt.net/CTranslate2/guides/transformers.html#bert) and in this issue are already using a classification head. Please create another issue if something is not working or missing.

I meant adding a classification head as part of ctranslate2; I would like to avoid the "transformers" package

**Comment by guillaumekln** [comment]
Please create a new issue. This is not related to the current issue.
